### PR TITLE
fix: standardize error and object serialization

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes N/A

### User facing changelog
n/a

### Additional details
After the large refactor changes incorporated in #20247, `switchToDomain` error stack invocations were temporarily removed until we could safely reincorporate them. This PR addresses re adding those stack invocations, as well as condolidating the error/thrown preprocessing logic with the config preprocessing logic.

 In `structuredClone`, Objects are serialized to literals while omitting prototype hierarchy values as well as omitting `get`, `set`, and other properties when cloning (see [here](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone)). 

`Error`s, even though they are objects, are handled differently depending on the implementation of `structuredClone`. Firefox cannot serialize `Error`s through `structuredClone` and chrome/polyfill implementations can clone `Error`s, but only keep `name`, `message`, and `stack`. These Error objects are also casted to object literals.

The changes here propose that we preprocess `Error`s and `Object`s the same before sending them off to `postMessage`. Regardless of the context, `Error`s and `Object`s now walk to prototype tree and add any serializable value to an object literal. This way, most serializable properties are preserved without having to deal with the caveats of `structuredClone`. Now that these are processed the same, a single method can be used to preprocess thrown errors as well as sync config/env.

Whenever a value is thrown from the secondary domain, we try to map it to an error. If this value cannot be mapped to an error to present to the user (serializable or otherwise), a `CypressError` is thrown with the message `'cy.switchToDomain()' could not serialize or map the thrown value. Please make sure the value being thrown is supported by the structured clone algorithm.`

A small system test was added to verify the stack trace behavior. This is likely a bit more appropriate for a `cy-in-cy` test, which we should convert once unification is merged. 

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by AtofStryker** [comment]
I am not sure we still need this [comment](https://github.com/cypress-io/cypress/blob/standardize_error_and_object_serialization/packages/driver/src/multi-domain/events/logs.ts#L16), unless the goal is to get invocation stacks from within the SWD callback?

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts [hunk 1]
```diff
@@ -240,21 +257,220 @@ describe('multi-domain - uncaught errors', { experimentalSessionSupport: true },
...
+      // @ts-ignore
+      cy.switchToDomain('foobar.com', () => {
+        throw Symbol('foo')
+      })
+    })
+
+    it('handles users throwing promises', (done) => {
+      cy.on('fail', (err) => {
+        expect(err.name).to.equal('CypressError')
+        expect(err.message).to.equal('`cy.switchToDomain()` could not serialize or map the thrown value. Please make sure the value being thrown is supported by the structured clone algorithm.')
```
Comment (AtofStryker) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:303]: I wonder if something like this makes more sense
```suggestion
        expect(err.message).to.equal('`cy.switchToDomain()` could not determine the thrown value. Please make sure the value being thrown is supported by the structured clone algorithm.')
```

Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:303]: I think just `serialize` is fine. Users will have to be familiar with the serialization limits when they're using the multi-domain API.

File: packages/driver/src/multi-domain/communicator.ts [hunk 2]
```diff
@@ -144,15 +101,26 @@ export class SpecBridgeDomainCommunicator extends EventEmitter {
   private windowReference
 
   private handleSubjectAndErr = (data: any = {}, send: (data: any) => void) => {
-    const { subject, err, ...rest } = data
+    let { subject, err, ...rest } = data
+
+    // check to see if the 'err' key is defined, and if it is, we have an error of any type
+    const hasMultiDomainError = !!Object.getOwnPropertyDescriptor(data, 'err')
```
Comment (chrisbreiding) [inline packages/driver/src/multi-domain/communicator.ts:107]: In what case does simply checking for `err` not work?

File: packages/driver/src/multi-domain/communicator.ts [hunk 3]
```diff
@@ -144,15 +101,26 @@ export class SpecBridgeDomainCommunicator extends EventEmitter {
...
+        try {
+          err = preprocessForSerialization(err)
+        } catch (e) {
+          err = e
+        }
+      }
+
       // We always want to make sure errors are posted, so clean it up to send.
-      send({ ...rest, subject, err: preprocessErrorForPostMessage(err) })
+      send({ ...rest, subject, err, hasMultiDomainError })
```
Comment (chrisbreiding) [inline packages/driver/src/multi-domain/communicator.ts:123]: Why is `hasMultiDomainError` necessary? I think we can just go off the existence of `err`, can't we?

File: packages/driver/src/cy/multi-domain/index.ts [hunk 4]
```diff
@@ -84,16 +87,12 @@ export function addCommands (Commands, Cypress: Cypress.Cypress, cy: Cypress.cy,
...
-
           reject(err)
         }
 
-        const onQueueFinished = ({ err, subject, unserializableSubjectType }) => {
-          if (err) {
-            return _reject(err)
+        const onQueueFinished = ({ err, subject, unserializableSubjectType, hasMultiDomainError }) => {
+          if (hasMultiDomainError) {
+            return _reject(reifyCrossDomainError(err, userInvocationStack))
```
Comment (chrisbreiding) [inline packages/driver/src/cy/multi-domain/index.ts:95]: Since the `switchToDomain` command itself is rejected, the userInvocationStack should be attached automatically. That's how it works for every other command. I'm curious why it's not working in some cases.

File: packages/driver/src/multi-domain/communicator.ts (hunk 2 above)
Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:107]: when `err` is `undefined` or falsy. [This](https://github.com/cypress-io/cypress/pull/20520/files#diff-24a3601e6f74e7b820b3d189b306a63af9acc391217e5e0318aff8be058279daR451) test case fails and hangs the test forever because it doesn't think an error has occured. Super rare, but does cover the falsy edge case.

File: packages/driver/src/multi-domain/communicator.ts (hunk 3 above)
Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:123]: Unfortunately not quite. I'm going to take a stab at addressing this comment [here](https://github.com/cypress-io/cypress/pull/20520#discussion_r823792751)

File: packages/driver/src/cy/multi-domain/index.ts (hunk 4 above)
Comment (AtofStryker) [reply packages/driver/src/cy/multi-domain/index.ts:95]: It looks like the `userInvocationStack` is only attached when the error is a chai assertion error, but commands errors do not contain `userInvocationStack`. From what I can see, command errors get the `userInvocationStack` from `current`. This works the same for `switchToDomain` as far as I can tell.

The issue here isn't so much with finding the `userInvocationStack` when the command is rejected, but with correcting the stack to come from `switchToDomain` as opposed to the `communicator` event emitter. We only need the `userInvocationStack` for this specifically, and once the stack is corrected, the `codeFrame` can be built properly in the `fail` handler inside of `cy.ts`.

##### without reify


##### with reify

Comment (AtofStryker) [reply packages/driver/src/cy/multi-domain/index.ts:95]: We could pull user invocation from `current`, but I feel like there are cases where the `switchToDomain` CB finishes, then a command fails in the secondary and `current` has moved on from `switchToDomain`. That's the only reason it is inside a closure.

Also right now we aren't reifying any cross origin uncaught errors, which I believe fails the current command. Since they are thrown on the spec window, I don't think we have to 😄

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts [hunk 5]
```diff
@@ -240,21 +257,220 @@ describe('multi-domain - uncaught errors', { experimentalSessionSupport: true },
...
+      cy.on('fail', (err: any) => {
+        expect(err.name).to.equal('CustomError')
+        expect(err.message).to.equal('custom error')
+        expect(err._name).to.equal('CustomError')
+        expect(err.foo).to.equal('bar')
 
-    cy.switchToDomain('foobar.com', () => {
-      cy.then(() => {
-        throw new class CustomError extends Error {
+        const { writable } = Object.getOwnPropertyDescriptor(err, 'name') || {}
```
Comment (mschile) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:322]:
```suggestion
        const { writable } = Object.getOwnPropertyDescriptor(err, 'name') as PropertyDescriptor
```

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts [hunk 6]
```diff
@@ -240,21 +257,220 @@ describe('multi-domain - uncaught errors', { experimentalSessionSupport: true },
...
+            this._name = name
+          }
+        }
+
+        const customErrorInstance = new CustomError('custom error')
+
+        // @ts-ignore
+        customErrorInstance.foo = 'bar'
+
+        const { writable } = Object.getOwnPropertyDescriptor(CustomError, 'name') || {}
```
Comment (mschile) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:346]:
```suggestion
        const { writable } = Object.getOwnPropertyDescriptor(CustomError, 'name') as PropertyDescriptor
```

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 1 above)
Comment (mschile) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:303]: @chrisbreiding, are you saying to remove `or map` and just have `serialize`?

File: packages/driver/src/util/serialization.ts [hunk 7]
```diff
@@ -27,12 +29,86 @@ const isSerializableInCurrentBrowser = (value: any) => {
...
+ * For instance:
+ * var a = new Error('myError')
+ * a.foo = 'bar'
+ * var b = structuredClone(a)
+ * b.foo // is undefined
+ *
+ * firefox
+ * structuredClone(new Error('myError')) -> throws error as native error cannot be serialized
+ *
+ * This method takes a similar the 'chromium' approach to structuredClone, except that the prototype chain is walked and ANY serializable value, including getters, is serialized.
```
Comment (mschile) [inline packages/driver/src/util/serialization.ts:82]:
```suggestion
 * This method takes a similar approach as the 'chromium' structuredClone algorithm, except that the prototype chain is walked and ANY serializable value, including getters, is serialized.
```

**Review by mschile — CHANGES_REQUESTED** [review]

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 6 above)
Comment (AtofStryker) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:346]: Since this is the test code and we have absolute control of `CustomError` I am good with this. Will update!

File: packages/driver/src/util/serialization.ts (hunk 7 above)
Comment (AtofStryker) [reply packages/driver/src/util/serialization.ts:82]: Thank you for finding this. This sentence doesn't make any sense 🥴 .

Comment (AtofStryker) [reply packages/driver/src/util/serialization.ts]: addressed in 4872216566f6ad7b26933bb99def80295f04e236

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 6 above)
Comment (AtofStryker) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts]: addressed in 4872216566f6ad7b26933bb99def80295f04e236

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 5 above)
Comment (AtofStryker) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts]: addressed in 4872216566f6ad7b26933bb99def80295f04e236

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 1 above)
Comment (chrisbreiding) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts:303]: Yeah, I'd say remove `or map`. I understand the intent behind it, but think that will just be confusing to users.

File: packages/driver/src/multi-domain/communicator.ts (hunk 2 above)
Comment (chrisbreiding) [reply packages/driver/src/multi-domain/communicator.ts:107]: I guess it just rubs me the wrong way to need this extra property for a couple rare edge cases when otherwise it's self-evident. How about we wrap the `err` in that case? Something like:

```js
// give the `err` truthiness if it's a falsey value like undefined/null/false
if (hasMultiDomainError && !err) {
  err = new Error(err)
}
```

Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:107]: I think as long as we cast the `err` to a string in the constructor, the behavior should be the same. Will update shortly!

Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:107]: addressed in f6c6975bfe

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts (hunk 1 above)
Comment (AtofStryker) [reply packages/driver/cypress/integration/e2e/multi-domain/multi_domain_uncaught_errors_spec.ts]: `or map` removed in c6648651f1

File: packages/driver/src/cy/multi-domain/index.ts [hunk 8]
```diff
@@ -84,16 +87,12 @@ export function addCommands (Commands, Cypress: Cypress.Cypress, cy: Cypress.cy,
...
           cleanup()
           log.error(err)
-          if (typeof err === 'object') {
-            err.onFail = () => {}
-          }
-
           reject(err)
         }
 
         const onQueueFinished = ({ err, subject, unserializableSubjectType }) => {
```
Comment (mjhenkes) [inline packages/driver/src/cy/multi-domain/index.ts:93]: What would you think of checking for the special object 'err' in the communicator's event listener and reifying the error there for any events that return the err keyword. If it's a concern that someone may return err, we could have a registry of events that this should be applied to... then you would have just one place to apply the reify logic (i have another event with an error i'd like to have this applied to.

https://github.com/cypress-io/cypress/blob/feature-multidomain/packages/driver/src/multi-domain/communicator.ts#L82

Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/index.ts:93]: I like this idea. Then it mirrors how it's handled on the spec bridge side.

Comment (AtofStryker) [reply packages/driver/src/cy/multi-domain/index.ts:93]: I like this idea, but the only thing I think that might be a problem is setting the correct stack. I am not sure how we can get the correct `userInvocationStack` in the `communicator` unless we use `current`. Any ideas?

Comment (chrisbreiding) [reply packages/driver/src/cy/multi-domain/index.ts:93]: We're already using `current` to the get the `userInvocationStack`. It's guaranteed to be `switchToDomain` because we don't resolve `switchToDomain` until either the secondary queue finishes or there's an error, in which case the queue is stopped for good. 

One option is we could pass the userInvocationStack into the spec bridge with `run:domain:fn` and replace the stack before sending it back, then we don't need to worry about it when reifying the error. This also addresses your question about [this comment](https://github.com/cypress-io/cypress/blob/standardize_error_and_object_serialization/packages/driver/src/multi-domain/events/logs.ts#L16). I think that may be necessary because the logs bypass `switchToDomain` and go straight to the runner. Though now that I think about it, I'm not sure if we want the log to have the `userInvocationStack`, the original stack, or a combination of both. But no need to worry about that for right now. Just pointing out that this solution (passing the userInvocationStack to the spec bridge) may be needed for that use case as well.

Comment (AtofStryker) [reply packages/driver/src/cy/multi-domain/index.ts:93]: OK I'm going to give this a shot. I'll pass `userInvocationStack` into `run:domain:fn` to set the invocation stack on the `primarytDomainCommunicator` and look for a special `err` key, and if its truthy, reify it.

**Review by mschile — APPROVED** [review]

File: packages/driver/src/multi-domain/communicator.ts [hunk 9]
```diff
@@ -117,6 +79,11 @@ export class PrimaryDomainCommunicator extends EventEmitter {
   }
 
   toSpecBridge (domain: string, event: string, data?: any) {
+    if (data?.userInvocationStack) {
+      this.userInvocationStack = data.userInvocationStack
+      delete data.userInvocationStack
+    }
```
Comment (mjhenkes) [inline packages/driver/src/multi-domain/communicator.ts:85]: Instead of attaching this as a byproduct of an event, why don't we attach it to the communicator explicitly. `PrimaryDomainCommunicator .setUserInvocationStack`

I think that would be a lot more clear.

**Review by mjhenkes — CHANGES_REQUESTED** [review]

File: packages/driver/src/multi-domain/communicator.ts (hunk 9 above)
Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:85]: @mjhenkes are you thinking calling `PrimaryDomainCommunicator.setUserInvocationStack` at the beginning of the `switchToDomain` call? Definitely would be a lot clearer as to when this happens and simpler

Comment (mjhenkes) [reply packages/driver/src/multi-domain/communicator.ts:85]: @AtofStryker yep, exactly!

Comment (AtofStryker) [reply packages/driver/src/multi-domain/communicator.ts:85]: I wonder if it makes sense to just set it on the instance directly, like `Cypress.multiDomainCommunicator.userInvocationStack = `state('current').get('userInvocationStack')`. I feel like this is super clear but wanted to get your thoughts

Comment (mjhenkes) [reply packages/driver/src/multi-domain/communicator.ts:85]: @AtofStryker I think that is fine too

**Review by mjhenkes — APPROVED** [review]
